pkg/cloud-init/cloud-init.go

```diff
@@ -47,24 +48,45 @@ const (
 )
 
 type DataSourceType string
+type DeviceMetadataType string
 
 const (
-	DataSourceNoCloud     DataSourceType = "noCloud"
-	DataSourceConfigDrive DataSourceType = "configDrive"
+	DataSourceNoCloud     DataSourceType     = "noCloud"
+	DataSourceConfigDrive DataSourceType     = "configDrive"
+	NICMetadataType       DeviceMetadataType = "nic"
+	DiskMetadataType      DeviceMetadataType = "disk"
 )
 
 // CloudInitData is a data source independent struct that
 // holds cloud-init user and network data
 type CloudInitData struct {
 	DataSource  DataSourceType
-	MetaData    string
+	MetaData    *Metadata
 	UserData    string
 	NetworkData string
+	DevicesData *[]DeviceData
+}
+
+type Metadata struct {
+	InstanceID    string        `json:"instance-id"`
+	LocalHostname string        `json:"local-hostname,omitempty"`
+	Hostname      string        `json:"hostname,omitempty"`
+	UUID          string        `json:"uuid,omitempty"`
+	Devices       *[]DeviceData `json:"devices,omitempty"`
+}
+
+type DeviceData struct {
+	Type    DeviceMetadataType `json:"type"`
+	Bus     string             `json:"bus"`
+	Address string             `json:"address"`
+	MAC     string             `json:"mac,omitempty"`
+	Serial  string             `json:"serial,omitempty"`
+	Tag     []string           `json:"tag"`
```

Comment:
should this be `tags`? At least `tags` is what I see in all examples.

Reply:
Yeah, I'll change it.